components/TARGET_PSA/TARGET_TFM/COMPONENT_NSPE/interface/src/tfm_ns_lock_rtx.c

```diff
@@ -53,6 +56,20 @@ uint32_t tfm_ns_lock_dispatch(veneer_fn fn,
         return TFM_ERROR_GENERIC;
     }
 
+    /* When the following conditions all assert, we can see it as we have 
+     * acquired NS lock and do dispatch:
+     * 1. Not in IRQ mode
+     * 2. IRQ masked
+     * 3. NS lock not owned by any thread */
+    bool can_dispatch_with_irq_masked = (!IsIrqMode()) && IsIrqMasked() && (!ns_lock.lock_count);
```

Comment:
In the case `!IsIrqMode() && IsIrqMasked() && ns_lock.lock_count`, you proceed to try to acquire the mutex, and crash because you can't acquire a mutex with interrupts disabled.

You should presumably return an error in that case.

Then, given the basic behaviour of this form has to be "try to do it if no-one else is", I'm not sure why it's worth messing with atomics and counters outside the mutex. Why not just set a flag inside the mutex?

    osMutexAcquire();
    ns_lock = true;
    fn();
    ns_lock = false;
    os_MutexRelease();

That will work fine.

Reply:
> Seems like a huge trap for developers - permitting secure calls with interrupts disabled that work 99.9% of the time, then start failing unexpectedly as the amount of secure-using threads increases.

Yes, it doesn't make sense. But during my NUMAKER_PFM_M2351 port, I find there's a code path which needs to call secure function in interrupt-disabled context:

<pre>
sleep_manager_sleep_auto > core_util_critical_section_enter > hal_sleep/hal_deepsleep > <b>CLK_Idle</b>/<b>CLK_PowerDown</b>
</pre>

Would you have solution to it?